compute_endpoint/globus_compute_endpoint/endpoint/endpoint_manager.py

```diff
@@ -761,6 +774,54 @@ def send_failure_notice(
         finally:
             sys.exit()
 
+    @contextmanager
+    def do_host_auth(self, username):
+        if not self._config.pam.enable:
+            pyprctl = _import_pyprctl()
+            yield
+
+            # If the administrator has *not* enabled PAM, then assume the
+            # intention is for a paranoid safe process and drop all
+            # privileges now ...
+            pyprctl.CapState().set_current()
+
+            # ... and stating that even if exec'ing might return some
+            # privileges, "no."  In particular after this, SETUID executables
+            # invoked from this process root will not get privileges
+            pyprctl.set_no_new_privs()
+
+            return
+
+        sname = self._config.pam.service_name
+        log.debug("PAM: Creating handle (%s, %s)", sname, username)
+        try:
+            PamHandle = _import_pamhandle()
+
+            with PamHandle(sname, username=username) as pamh:
+                log.debug("PAM: Invoking account stage")
+                pamh.pam_acct_mgmt()
+                log.debug("PAM: Creating credentials")
+                pamh.credentials_establish()
+                log.debug("PAM: Opening session")
+                pamh.pam_open_session()
+
+                yield
+
+                # wiped by initgroups, so reinitialize
+                log.debug("PAM: Recreating credentials")
+                pamh.credentials_establish()
+                log.debug("PAM: Closing session")
+                pamh.pam_close_session()
+                log.debug("PAM: Removing credentials")
+                pamh.credentials_delete()
+
+                log.debug("PAM: Closing handle")
+        except Exception as e:
+            log.error(str(e))  # Share (very likely) pamlib error with admin ...
+
+            # ... but be opaque with user.
+            raise PermissionError("see your system administrator") from None
```

Comment:
Given this will catch errors import or initializing pamlib, it might be good to give a more useful error in that case since that is a fundamentally config error?  Or perhaps even try and catch it somewhere _before_ the attempt start the UEP, e.g. on MEP start up if config.pam.enabled then try import first?

Reply:
As regard the "useful error in that case," we currently do have this import error if we can't even find PAM:
```python
pam_lib_name = find_library("pam")
if pam_lib_name is None:
    msg = (
        "Failed to find `pam` shared library\n"
        "\n  Hints:"
        "\n    - install libpam.so? (via yum, apt, etc.)"
        "\n    - ldconfig -p"
        "\n    - LD_PRELOAD"
    )
    raise ImportError(msg)
```
... which would be emitted to the logs per the `.error()` call there.

But failing at startup on a bad import is a much better UX.  Good suggestion.  I'll look at that.

Reply:
Have just pushed a commit based on this feedback.

Reply:
Commit does two things:
- Attempt the import during initialization (`__init__`)
- Catch specifically the expected `pam.PamError` and handle that gracefully, while also catching all other exceptions with an exception-in-the-log explosion.

The larger-than-expected changes in the tests are me recognizing an opportunity for pytest fixtures that I missed in my first go-round of test implementation.  The new tests in this commit are:
- `test_conditional_imports_verified_at_init_for_ux`
- Additional parameter `exc` in `test_pam_error`